vital/types/object_track_set.cxx

```diff
@@ -44,29 +44,61 @@ namespace vital {
 
 typedef std::unique_ptr< track_set_implementation > tsi_uptr;
 
-
 /// Default Constructor
 object_track_set
 ::object_track_set()
   : track_set( tsi_uptr( new simple_track_set_implementation ) )
 {
 }
 
-
 /// Constructor specifying the implementation
 object_track_set
 ::object_track_set( std::unique_ptr<track_set_implementation> impl )
   : track_set( std::move( impl ) )
 {
 }
 
-
 /// Constructor from a vector of tracks
 object_track_set
 ::object_track_set( std::vector< track_sptr > const& tracks )
   : track_set( tsi_uptr( new simple_track_set_implementation( tracks ) ) )
 {
 }
 
+detected_object_sptr& object_track_state::detection()
+{
+  return detection_;
+}
+const detected_object_cptr object_track_state::detection() const
+{
+  return std::const_pointer_cast<const detected_object>(detection_);
```

Comment:
Please remove all of these unnecessary `const_cast`s. A `const_cast` signals that something potentially dangerous is happening. These are unnecessary and add noise for someone checking for potentially dangerous casts.

(Note: This point specifically is the reason this review is Request Changes rather than Comment.)

Reply:
Should kwiver support having a const access to on objects that encapsulate other objects?
If the object is const, I would assume the encapsulated object should not be changed either and also be a const object. 

That is what I was trying to do here, but the use of const_cast does not look to be correct.
Perhaps, if we want to support the above assumption, I should return a const* to the raw object

Reply:
> Perhaps [...] I should return a const* to the raw object

Not necessary; conversion from `std::shared_ptr<T>` to `std::shared_ptr<T const>` is implicit. The cast is superfluous.

That said, I'm not strongly opinionated on going through the extra gyrations to return `const` sub-objects when the parent object is `const`. The benefit is obvious, but it's extra work and I think we don't usually bother. What's more important is either properly deep copying the sub-objects in `clone()` (see #812) or making the sub-objects *always* immutable. (#812 modifies e.g. the image mask such that you can never get a mutable image mask; you can only change a detection's image mask by replacing the existing mask with a new instance. This makes it safe to share the image mask even through "deep" copies; it is essentially an explicit copy-on-write mechanism.)

Reply:
I get back a null std::shared_ptr<T const> if i do not have the const_pointer_cast

Reply:
Really? It works for me: https://godbolt.org/z/Q3C7pL.

Reply:
nevermind, not sure what I was seeing before

Reply:
So... now there is a `const_pointer_cast`? I believe that is unneeded? (Especially as this seems to be the only instance where it is present...)